src/operators/join.rs

```diff
@@ -320,100 +320,127 @@ impl<G, T1> JoinCore<G, T1::Key, T1::Val, T1::R> for Arranged<G,T1>
-                                    // A trace should provide the contract that whatever its `set_physical_compaction` capability,
-                                    // it is safe (and reasonable) to await delivery of batches up through that frontier.
-                                    // In this case, we should be able to await (not block on) the arrival of these batches.
+                            // Ignore any pre-loaded data.
+                            if PartialOrder::less_equal(&acknowledged1, &batch1.lower()) {
+                                if !batch1.is_empty() {
+                                    // It is safe to ask for `ack2` as we validated that it was at least `get_physical_compaction()`
+                                    // at start-up, and have held back physical compaction ever since.
                                     let (trace2_cursor, trace2_storage) = trace2.cursor_through(acknowledged2.borrow()).unwrap();
                                     let batch1_cursor = batch1.cursor();
-                                    todo1.push_back(Deferred::new(trace2_cursor, trace2_storage, batch1_cursor, batch1.clone(), capability.clone(), |r2,r1| (r1.clone()) * (r2.clone())));
+                                    todo1.push_back(Deferred::new(trace2_cursor, trace2_storage, batch1_cursor, batch1.clone(), capability.clone()));
                                 }
-                            }
 
-                            // It would be alarming (incorrect) to receieve a non-empty batch that does not advance the
-                            // acknowledged frontier, as each batch must be greater than previous batches, and the input.
-                            // Empty batches may be received as information races between frontier forwarding of the trace
-                            // and the empty batches themselves (which can be sent as part of trace importing).
-                            if acknowledged1.is_none() { acknowledged1 = Some(timely::progress::frontier::Antichain::from_elem(<G::Timestamp>::minimum())); }
-                            if let Some(acknowledged1) = &mut acknowledged1 {
-                                if !PartialOrder::less_equal(&*acknowledged1, batch1.upper()) {
-                                        if !batch1.is_empty() {
-                                        panic!("Non-empty batch1 upper not beyond acknowledged frontier: {:?}, {:?}", batch1.upper(), acknowledged1);
-                                    }
-                                }
+                                // To update `acknowledged1` we might presume that `batch1.lower` should equal it, but we
+                                // may have skipped over empty batches. Still, the batches are in-order, and we should be
+                                // able to just assume the most recent `batch1.upper`
                                 acknowledged1.clone_from(batch1.upper());
```

Comment:
should we maybe assert that acknowledged1 <= batch1.upper()? I feel like everything hinges on that (and we assert in `set_physical_compaction` so maybe its fine)

Reply:
We could! There would be a tremendous error if we minted batches that were not in-order, increasing. I'll put it in as a `debug_assert`; it's one of these things that has some run-time cost and should be `assert`ed elsewhere in the code.